win: enumerate drives via FindNextVolumeW + GetVolumePathNamesForVolumeNameW

chklauser opened the pull request:
This PR changes the mechanism for enumerating "disks" on windows. The previous approach only worked for volumes that were assigned a drive letter. Windows volumes can also be mounted in the file system.

We first fetch a list of volume GUID paths (volume names). These have the form `\\?\{xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx}\`. We then use `GetVolumePathNamesForVolumeNameW` to map each of these to a set of mount paths (`C:\` for drive letters or arbitrary paths for path mounts). We then return one `Disk` entry for each mount path (flat-map).

This should roughly match the semantics of the linux implementation, which also primarily looks at mount points (not actual disk volumes). It means that a volume that is mounted multiple times (multiple drive letters and/or mount paths) will show up as multiple disks.

The output of the simple example changes as follows:
Before:
```
> disks
Disk("")[FS: ['N', 'T', 'F', 'S']][Type: SSD][removable: no] mounted on "C:\\": 6045496340480/8000072638464 B
Disk("User")[FS: ['N', 'T', 'F', 'S']][Type: HDD][removable: no] mounted on "F:\\": 724046233600/3000457228288 B
Disk("Quick")[FS: ['N', 'T', 'F', 'S']][Type: SSD][removable: no] mounted on "Q:\\": 262496968704/2000381014016 B
```
After:
```
> disks
Disk("Hot")[FS: ['N', 'T', 'F', 'S']][Type: SSD][removable: no] mounted on "C:\\Hot\\": 606733127680/2048390066176 B
Disk("User")[FS: ['N', 'T', 'F', 'S']][Type: HDD][removable: no] mounted on "F:\\": 724046233600/3000457228288 B
Disk("User")[FS: ['N', 'T', 'F', 'S']][Type: HDD][removable: no] mounted on "C:\\Cold\\": 724046233600/3000457228288 B
Disk("")[FS: ['N', 'T', 'F', 'S']][Type: SSD][removable: no] mounted on "C:\\": 6043730104320/8000072638464 B
Disk("Quick")[FS: ['N', 'T', 'F', 'S']][Type: SSD][removable: no] mounted on "Q:\\": 262496968704/2000381014016 B
```

I did check the impact on `cargo bench` on my machine
```
origin/master:
test bench_refresh_disks_list      ... bench:     129,725 ns/iter (+/- 5,296)

windows-disks-mount-path:
test bench_refresh_disks_list      ... bench:     250,926 ns/iter (+/- 9,357)
```

I did a quick trace with the Windows Performance Recorder and it seems to me like all the time gets spent in `GetVolumePathNamesForVolumeNameW`. Seems like that's the price to pay here. Below is a screenshot of one of the bench threads. 
![image](https://github.com/GuillaumeGomez/sysinfo/assets/149009/77d3b9ea-f034-4913-ab59-c94ff6340ffc)

# Contribution
I noticed that the Nushell [`sys` command](https://github.com/nushell/nushell/blob/d0dc6986dd14e7ea5443f3dc0eb5b101c5de5efc/crates/nu-command/src/system/sys.rs#L108) did not list mount paths and traced the omission to this crate. It seemed like a fine excuse to write some "low level" systems code and try out the `windows` crate.

I do realize that this is 100% a drive-by PR and that the feature is _relatively_ exotic. If you don't think the PR is a good fit for the crate, please do feel free to close it without comment 😄 . I learned a lot in the process and getting the PR merged would just be a little bonus on top. That said, if you _are_ interested in the PR, I'm happy to work on it 😄

GuillaumeGomez commented:
That looks super cool and the explanations are super detailed, thanks! I'll review it tomorrow as it's a bit late for me currently.

chklauser commented on src/windows/disk.rs:
```diff
@@ -179,40 +271,36 @@ pub(crate) unsafe fn get_list() -> Vec<Disk> {
-                pos += 1;
+
+            let mount_paths = get_volume_path_names_for_volume_name(&volume_name[..]);
+            if !mount_paths.is_empty() {
```
```suggestion
            if mount_paths.is_empty() {
```

GuillaumeGomez commented on src/windows/disk.rs:
```diff
@@ -19,12 +23,110 @@ use windows::Win32::System::Ioctl::{
+            }
+            volume_names.push(from_zero_terminated(&buf[..]));
+        }
+        if FindVolumeClose(handle) != Ok(()) {
```
```suggestion
        if FindVolumeClose(handle).is_err() {
```

GuillaumeGomez commented on src/windows/disk.rs:
```diff
@@ -19,12 +23,110 @@ use windows::Win32::System::Ioctl::{
+    let mut path_names_buf = vec![0u16; MAX_PATH as usize];
+    let mut path_names_output_size = 0u32;
+    for _ in 0..10 {
+        match GetVolumePathNamesForVolumeNameW(
+            volume_guid_path,
+            Some(path_names_buf.as_mut_slice()),
+            &mut path_names_output_size)
+            .map_err(|_| GetLastError()
+                .expect_err("GetLastError should return an error after GetVolumePathNamesForVolumeNameW returned zero.")
+                .code()) {
```
The code is quite difficult to read as is. What do you think about (you will need to run `cargo fmt` on it ^^'):

```suggestion
    let volume_path_names = GetVolumePathNamesForVolumeNameW(
        volume_guid_path,
        Some(path_names_buf.as_mut_slice()),
        &mut path_names_output_size);
    let code = volume_path_names
        .map_err(|_| match GetLastError() {
            Ok(_) => {
                sysinfo_debug!("GetLastError should return an error after GetVolumePathNamesForVolumeNameW returned zero.");
                // We return an error in any case that is not `ERROR_MORE_DATA` to stop the 
                ERROR_UNRECOGNIZED_VOLUME
            }
            Err(e) => e.code(),
        });
        match code {
```

?

I also changed how `GetLastError` error code is handled to ensure it can never panic.

chklauser replied:
Yep, your version is more readable. I was leaning heavily on `let-else` in an earlier version of the code. It makes these manual error condition checks _much_ easier to read 😉. But that would require MSRV 1.65 🥲

As for the `GetLastError` I agree that we should probably not panic. At first, I was really confused by `GetLastError` returning a `Result`, but [according to Microsoft](https://learn.microsoft.com/en-us/windows/win32/api/errhandlingapi/nf-errhandlingapi-getlasterror#return-value), there _are_ indeed functions that have you _always_ call `GetLastError()`, which then _sometimes_ returns `0` to indicate success ¯\_(ツ)_/¯

GuillaumeGomez replied:
> But that would require MSRV 1.65 🥲

I didn't know it was stable. I don't mind updating the MSRV to 1.65 but let's do it in a follow-up PR to prevent changing to many thing in this one. ;)

> As for the `GetLastError` I agree that we should probably not panic. At first, I was really confused by `GetLastError` returning a `Result`, but [according to Microsoft](https://learn.microsoft.com/en-us/windows/win32/api/errhandlingapi/nf-errhandlingapi-getlasterror#return-value), there _are_ indeed functions that have you _always_ call `GetLastError()`, which then _sometimes_ returns `0` to indicate success ¯\\\_(ツ)\_/¯

Oh well, windows API is weird, nothing new. ^^'

GuillaumeGomez commented on src/windows/disk.rs:
```diff
@@ -144,33 +246,23 @@ unsafe fn get_drive_size(mount_point: &[u16]) -> Option<(u64, u64)> {
 
             if drive_type != DRIVE_FIXED && drive_type != DRIVE_REMOVABLE {
-                return None;
+                return vec![];
```
For consistency, let's use `Vec::new()`.

GuillaumeGomez commented on src/windows/disk.rs:
```diff
@@ -179,40 +271,36 @@ pub(crate) unsafe fn get_list() -> Vec<Disk> {
             if !volume_info_res {
-                return None;
+                sysinfo_debug!("Error: GetVolumeInformationW = {:?}", GetLastError());
+                return vec![];
```
`Vec::new()`.

GuillaumeGomez commented on src/windows/disk.rs:
```diff
@@ -179,40 +271,36 @@ pub(crate) unsafe fn get_list() -> Vec<Disk> {
+            let handle = match HandleWrapper::new(&device_path[..], Default::default()) {
+                Some(h) => h,
+                None => {
+                    return vec![];
```
`Vec::new()`.

GuillaumeGomez commented on src/windows/disk.rs:
```diff
@@ -179,40 +271,36 @@ pub(crate) unsafe fn get_list() -> Vec<Disk> {
+            let (total_space, available_space) = match get_drive_size(&mount_paths[0][..]) {
+                Some(space) => space,
+                None => {
+                    return vec![];
```
`Vec::new()`.

GuillaumeGomez commented on src/windows/disk.rs:
```diff
@@ -179,40 +271,36 @@ pub(crate) unsafe fn get_list() -> Vec<Disk> {
             if total_space == 0 {
-                return None;
+                sysinfo_debug!("total_space == 0");
+                return vec![];
```
`Vec::new()`.

GuillaumeGomez commented on src/windows/disk.rs:
```diff
@@ -245,18 +333,29 @@ pub(crate) unsafe fn get_list() -> Vec<Disk> {
+            let file_system = file_system
+                .iter()
+                .take_while(|c| **c != 0)
+                .map(|c| *c as u8)
```
Is there a case where a `u16` could be needed here?

chklauser replied:
You tell me 😉! The public [`Disk` interface](https://github.com/GuillaumeGomez/sysinfo/blob/ade89e39c26599db39f2a4a847e79908fbab1154/src/common.rs#L1925C5-L1925C39) exposes a `&[u8]`. The [original code](https://github.com/GuillaumeGomez/sysinfo/blob/ade89e39c26599db39f2a4a847e79908fbab1154/src/windows/disk.rs#L201C43-L201C94) also truncates the code units to `u8` 😢 

My gut reaction is that I would make the `Disk` interface expose a `&OsStr` instead, but that seems like a big hassle (breaking change) for relatively little benefit.

GuillaumeGomez replied:
Next release will be a major one as I already completely changed the API. If you don't mind changing that (either in this PR or in a follow-up), I'll gladly accept it.

chklauser replied:
Happy to do the `[u8]` -> `OsStr` change, but I'd prefer to do it in a separate PR.

GuillaumeGomez replied:
I'd prefer too so let's go for another PR.

GuillaumeGomez commented:
Mostly nits at this point. Otherwise looks good to me!

chklauser commented:
Thank you for the review! I think I have addressed the comments. Can you re-try the one check that failed? (I don't seem to have the permission) I think it just didn't manage to download something (not a windows target).

GuillaumeGomez commented:
> Thank you for the review! I think I have addressed the comments. Can you re-try the one check that failed? (I don't seem to have the permission) I think it just didn't manage to download something (not a windows target).

It's not an issue. I restarted just in case but it's not coming from your changes.

GuillaumeGomez commented on src/windows/disk.rs:
```diff
@@ -19,12 +24,118 @@ use windows::Win32::System::Ioctl::{
+            .map_err(|_| match GetLastError() {
+                Ok(_) => {
+                    sysinfo_debug!("GetLastError should return an error after GetVolumePathNamesForVolumeNameW returned zero.");
+                    // We return an error in any case that is not `ERROR_MORE_DATA` to stop the
```
Missing word at the end of the sentence. :wink:

GuillaumeGomez reviewed (approved):
Looks all good now, thanks! Do you want to send a PR for `let else` (which would increase the MSRV) and the switch to `OsStr` instead of `Vec<u8>`?

chklauser commented:
(Didn't see your latest comment)

Happy to do the `[u8]` -> `OsStr` change. I'd prefer to do it in a separate PR. 

I can also up the MSRV (1.65.0) and use let-else in `windows/disk.rs`, but I wouldn't go hunting across an unfamiliar code base to sprinkle `let-else`s everywhere.

GuillaumeGomez commented:
> I can also up the MSRV (1.65.0) and use let-else in `windows/disk.rs`, but I wouldn't go hunting across an unfamiliar code base to sprinkle `let-else`s everywhere.

It's fine, just update your code, doesn't need to go through the whole codebase.
